internal/api/impl/v1/user/service.go

```diff
@@ -0,0 +1,123 @@
+	}
+	return nil, fmt.Errorf("%w: wrong entity format, attempting user format, received '%T'", shared.BadRequestError, entity)
+}
+
+func (s *service) create(entity *v1.User) (*v1.PublicUser, error) {
+	// Update the time contains in the entity
+	entity.Metadata.CreateNow()
+	// check that the password is correctly filled
+	if len(entity.Spec.Password) == 0 {
+		return nil, fmt.Errorf("%w: password cannot be empty", shared.BadRequestError)
+	}
+	hash, err := hashAndSalt(entity.Spec.Password)
+	if err != nil {
+		logrus.WithError(err).Errorf("unable to generate the hash for the password of the user %s", entity.Metadata.Name)
+		return nil, shared.InternalError
+	}
+	// save the hash in the password field
+	entity.Spec.Password = hash
+	if createErr := s.dao.Create(entity); createErr != nil {
+		return nil, createErr
+	}
+	return v1.NewPublicUser(entity), nil
+}
+
+func (s *service) Update(entity api.Entity, parameters shared.Parameters) (interface{}, error) {
+	if userObject, ok := entity.(*v1.User); ok {
+		return s.update(userObject, parameters)
+	}
+	return nil, fmt.Errorf("%w: wrong entity format, attempting user format, received '%T'", shared.BadRequestError, entity)
+}
+
+func (s *service) update(entity *v1.User, parameters shared.Parameters) (*v1.PublicUser, error) {
+	if entity.Metadata.Name != parameters.Name {
+		logrus.Debugf("name in user '%s' and coming from the http request: '%s' doesn't match", entity.Metadata.Name, parameters.Name)
+		return nil, fmt.Errorf("%w: metadata.name and the name in the http path request doesn't match", shared.BadRequestError)
+	}
+	// find the previous version of the project
+	oldEntity, err := s.dao.Get(parameters.Name)
+	if err != nil {
+		return nil, err
+	}
+	entity.Metadata.Update(oldEntity.Metadata)
+	// in case the user updated his password, then we should hash it again, otherwise the old password should be kept
+	if len(entity.Spec.Password) > 0 {
+		hash, hashErr := hashAndSalt(entity.Spec.Password)
+		if hashErr != nil {
+			logrus.WithError(hashErr).Errorf("unable to generate the hash for the password of the user %q", entity.Metadata.Name)
+			return nil, hashErr
+		}
+		entity.Spec.Password = hash
+	} else {
+		entity.Spec.Password = oldEntity.Spec.Password
+	}
+	// in case the user is updating the firstname / lastname, then it should be updated, otherwise the old one should be kept
+	if len(entity.Spec.FirstName) == 0 {
+		entity.Spec.FirstName = oldEntity.Spec.FirstName
+	}
+	if len(entity.Spec.LastName) == 0 {
+		entity.Spec.LastName = oldEntity.Spec.LastName
+	}
+	if updateErr := s.dao.Update(entity); updateErr != nil {
+		logrus.WithError(err).Errorf("unable to perform the update of the user %q", entity.Metadata.Name)
+		return nil, updateErr
+	}
+	return v1.NewPublicUser(entity), nil
+}
+
+func (s *service) Delete(parameters shared.Parameters) error {
+	return s.dao.Delete(parameters.Name)
+}
+
+func (s *service) Get(parameters shared.Parameters) (interface{}, error) {
```

Comment:
The dao return a User. Does it mean that the get will return the password ?

Reply:
ah yeah I miss this part. That wasn't intended. I will fix that